R/get-based-on.R

```diff
@@ -0,0 +1,169 @@
+    }
+  )
+}
+
+
+#' @rdname get_based_on
+#' @param .bbi_object Character scaler of a path to a model that can be loaded with `read_model(.bbi_object)`
+#' @export
+get_based_on.character <- function(.bbi_object, .check_exists = FALSE) {
+
+  .bbi_object <- tryCatch(
+    {
+      read_model(.bbi_object)
+    },
+    error = function(e) {
+      stop(glue("Cannot load model object from path `{.bbi_object}` :\n{paste(e, collapse = '\n')}"))
+    }
+  )
+
+  return(get_based_on(.bbi_object, .check_exists = .check_exists))
+}
+
+
+#' @rdname get_based_on
+#' @param .bbi_object Tibble of class `bbi_run_log_df`
+#' @importFrom purrr map
+#' @export
+get_based_on.bbi_run_log_df <- function(.bbi_object, .check_exists = FALSE) {
+
+  .out_paths <- map(.bbi_object[[ABS_MOD_PATH]], function(.path) {
+    get_based_on(.path, .check_exists = .check_exists)
+  })
+
+  return(.out_paths)
+}
+
+
+#' Get model ancestry
+#'
+#' Extract paths to all models that this model is based on (iterating over `get_based_on()` function),
+#' and all models that those models are based on, recursively.
+#' Returns a sorted unique character vector.
+#' @param .bbi_object The model object (or path, etc.) to query
+#' @rdname get_based_on
+#' @export
+get_model_ancestry <- function(.bbi_object) {
+  UseMethod("get_model_ancestry")
+}
+
+#' The default method attempts to extract the path from any object passed to it
+#' @importFrom purrr map
+#' @importFrom stringr str_detect
+#' @param .bbi_object The object to attempt to query. Could be a partially built bbi_{.model_type}_object or some other custom object containing model data.
+#' @rdname get_based_on
+#' @export
+get_model_ancestry.default <- function(.bbi_object) {
+  .checked <- c()
+  .to_check <- get_based_on(.bbi_object)
+  .results <- .to_check
+
+  while(length(.to_check) > 0) {
+    # record this round of models as being checked
+    .checked <- c(.checked, .to_check)
+
+    # get based_on for this round of models
+    .this_res <- map(.to_check, function(.p) {
+      .res <- tryCatch(
+        {
+          get_based_on(.p)
+        },
+        error = function(e) {
+          if (str_detect(e$message, "read_model.+found no YAML")) {
+            stop(glue("Found {.p} in get_model_ancestry() tree, but could not find a YAML file for that model."), call. = FALSE)
+          }
+          stop(e$message)
+        }
+      )
+
+      return(.res)
+    })
+    .this_res <- unique(unlist(.this_res))
+
+    # add to results
+    .results <- unique(c(.results, .this_res))
+
+    # see if there are any we haven't checked yet
+    .to_check <- .this_res[!(.this_res %in% .checked)]
+  }
+
+  return(sort(.results))
+}
+
+
+#' @rdname get_based_on
+#' @param .bbi_object Character scaler of a path to a model that can be loaded with `read_model(.bbi_object)`
+#' @export
+get_model_ancestry.character <- function(.bbi_object) {
```

Comment:
think about memoise package

Reply:
Looked at `memoise` but I don't actually want to do this because the same input would have different results if the underlying yaml's have changed. Keeping track of that would get complicated quickly.

Did a refactor that sped things up dramatically. With the code below, it runs in half a second on 300 models, each with 3 different models in their `based_on` field.

```

MODS_TO_TEST <- 300
mod_names <- as.character(seq_len(MODS_TO_TEST))

mods <- map(mod_names, function(.m) {
  suppressSpecificWarning(new_model(yaml_ext(.m), .description = "fake"), .regexpr = "No model file found")
})

BASED_ON_COUNT <- 3
mods <- map(mods, ~ replace_based_on(.x, sample(mod_names, BASED_ON_COUNT)))


log_df <- run_log()
# View(log_df)

bo_bm <- rbind(
  benchmark(based_on_list <- log_df %>% get_based_on(), replications = 1),
  benchmark(based_on_list <- log_df %>% get_based_on(), replications = 10)
)
bo_bm
#                                        test replications elapsed relative user.self sys.self user.child sys.child
# 1 based_on_list <- log_df %>% get_based_on()            1   0.449        1     0.417    0.032          0         0
# 2 based_on_list <- log_df %>% get_based_on()           10   5.293        1     4.949    0.310          0         0

anc_bm <- rbind(
  benchmark(ancestry_list <- log_df %>% get_model_ancestry(), replications = 1),
  benchmark(ancestry_list <- log_df %>% get_model_ancestry(), replications = 10)
)
anc_bm
#                                               test replications elapsed relative user.self sys.self user.child sys.child
# 1 ancestry_list <- log_df %>% get_model_ancestry()            1   1.636        1     1.613    0.024          0         0
# 2 ancestry_list <- log_df %>% get_model_ancestry()           10  17.382        1    17.045    0.339          0         0
```